add testing support for Vectorized<Half>

metascroy opened the pull request:
Summary: Implement half neon vectorization.

Test Plan:
Adjusted fbcode//caffe2:vec_test_all_types so that Half type is covered by tests.  On x86, this just tests the half type template in vec_base.h.  All tests pass.

```
buck2 test fbcode//caffe2:vec_test_all_types
Buck UI: https://www.internalfb.com/buck2/1062c33e-3e2b-4915-929d-5717f4d05ea8
Test UI: https://www.internalfb.com/intern/testinfra/testrun/15762598722963484
Network: Up: 7.6MiB  Down: 211MiB  (reSessionID-f19e6afc-3fa5-4455-87c5-df6f64d2ef70)
Jobs completed: 2872. Time elapsed: 8:17.6s.
Cache hits: 16%. Commands: 1571 (cached: 247, remote: 1321, local: 3)
Tests finished: Pass 336. Fail 0. Fatal 0. Skip 0. Build failure 0
```

On M1 mac, I built pytorch from source with
```
cd pytorch
python setup.py develop
```

and then executed

```
 .build/bin/vec_test_all_types_DEFAULT
```

This actually tests vec256_half_neon.h and results in
```
[==========] 336 tests from 139 test suites ran. (6492 ms total)
[  PASSED  ] 334 tests.
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] SqrtAndReciprocalReal/2.RSqrt, where TypeParam = at::vec::DEFAULT::Vectorized<c10::Half>
[  FAILED  ] SqrtAndReciprocalReal/2.Reciprocal, where TypeParam = at::vec::DEFAULT::Vectorized<c10::Half>
```

The 2 failing test cases are not failing bad (e.g., 0.000238776 != 0.000238895), I just need to play around with the tolerances a bit.

Differential Revision: D55226003


cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

malfet commented on aten/src/ATen/native/Math.h:
```diff
@@ -147,7 +148,7 @@ jiterator_also_stringify_as(jiterator_code(
 
 template <typename T>
-static inline typename std::enable_if<std::is_floating_point<T>::value, T>::type
+static inline typename std::enable_if<std::is_floating_point<T>::value || std::is_reduced_floating_point_v<T>, T>::type
```
Nit
```suggestion
static inline typename std::enable_if_t<std::is_floating_point_v<T>|| std::is_reduced_floating_point_v<T>, T>
```

malfet commented on aten/src/ATen/test/vec_test_all_types.cpp:
```diff
@@ -1,5 +1,6 @@
 #include <ATen/test/vec_test_all_types.h>
 #include <c10/util/irange.h>
+#include "c10/util/Half.h"
```
```suggestion
#include <c10/util/Half.h>
```

malfet commented on aten/src/ATen/test/vec_test_all_types.h:
```diff
@@ -1447,7 +1464,7 @@ template<typename T>
     using UVT = UvalueType<T>;
     TestingCase<T> testCase;
-    if (!bitwise && std::is_floating_point<UVT>::value) {
+    if (!bitwise && (std::is_floating_point<UVT>::value || std::is_reduced_floating_point_v<UVT>)) {
```
```suggestion
    if (!bitwise && (std::is_floating_point_v<UVT> || std::is_reduced_floating_point_v<UVT>)) {
```

malfet commented on aten/src/ATen/native/Math.h:
```diff
@@ -12,6 +12,7 @@
 #include <cstdlib>
 #include <limits>
 #include <type_traits>
+#include "c10/util/BFloat16-math.h"
```
This shouldn't be needed, should it?
```suggestion
```

malfet commented on aten/src/ATen/test/vec_test_all_types.h:
```diff
@@ -420,15 +425,15 @@ void filter_clamp(T& f, T& s, T& t) {
 
 template <typename T>
-std::enable_if_t<std::is_floating_point<T>::value, void> filter_fmadd(T& a, T& b, T& c) {
+std::enable_if_t<std::is_floating_point<T>::value || std::is_reduced_floating_point_v<T>, void> filter_fmadd(T& a, T& b, T& c) {
```
```suggestion
std::enable_if_t<std::is_floating_point_v<T> || std::is_reduced_floating_point_v<T>, void> filter_fmadd(T& a, T& b, T& c) {
```

malfet commented on aten/src/ATen/test/vec_test_all_types.h:
```diff
@@ -420,15 +425,15 @@ void filter_clamp(T& f, T& s, T& t) {
 
 template <typename T>
-std::enable_if_t<std::is_floating_point<T>::value, void> filter_fmod(T& a, T& b) {
+std::enable_if_t<std::is_floating_point<T>::value || std::is_reduced_floating_point_v<T>, void> filter_fmod(T& a, T& b) {
```
```suggestion
std::enable_if_t<std::is_floating_point_v<T> || std::is_reduced_floating_point_v<T>, void> filter_fmod(T& a, T& b) {
```

malfet commented on aten/src/ATen/test/vec_test_all_types.h:
```diff
@@ -346,7 +351,7 @@ nearlyEqual(T a, T b, T tolerance) {
 
 template<class T>
-std::enable_if_t<!std::is_floating_point<T>::value, bool>
+std::enable_if_t<!(std::is_floating_point<T>::value || std::is_reduced_floating_point_v<T>), bool>
```
```suggestion
std::enable_if_t<!(std::is_floating_point_v<T> || std::is_reduced_floating_point_v<T>), bool>
```

malfet commented on aten/src/ATen/test/vec_test_all_types.h:
```diff
@@ -330,7 +335,7 @@ T safe_fpt_division(T f1, T f2)
 
 template<class T>
-std::enable_if_t<std::is_floating_point<T>::value, bool>
+std::enable_if_t<std::is_floating_point<T>::value || std::is_reduced_floating_point_v<T>, bool>
```
```suggestion
std::enable_if_t<std::is_floating_point_v<T>|| std::is_reduced_floating_point_v<T>, bool>
```

malfet commented on aten/src/ATen/test/vec_test_all_types.h:
```diff
@@ -268,38 +269,42 @@ std::ostream& operator<<(std::ostream& stream, const CheckWithinDomains<T>& dmn)
-    T cmax = std::is_same<T, float>::value ? static_cast<T>(1e+30) : static_cast<T>(1e+300);
-    T cmin = std::is_same<T, float>::value ? static_cast<T>(-1e+30) : static_cast<T>(-1e+300);
+std::enable_if_t<std::is_floating_point<T>::value || std::is_reduced_floating_point_v<T>, bool> check_both_big(T x, T y) {
+    T cmax = std::is_same<T, c10::Half>::value ? static_cast<T>(5e+4) : (
+        std::is_same<T, float>::value ? static_cast<T>(1e+30) : static_cast<T>(1e+300)
+    );
+    T cmin = std::is_same<T, c10::Half>::value ? static_cast<T>(-5e+4) : (
+        std::is_same<T, float>::value ? static_cast<T>(-1e+30) : static_cast<T>(-1e+300)
+    );
```
```suggestion
    T cmax = std::is_same_v<T, c10::Half> ? static_cast<T>(5e+4) : (
        std::is_same_v<T, float>? static_cast<T>(1e+30) : static_cast<T>(1e+300)
    );
    T cmin = std::is_same_v<T, c10::Half>? static_cast<T>(-5e+4) : (
        std::is_same_v<T, float>? static_cast<T>(-1e+30) : static_cast<T>(-1e+300)
    );
```

malfet commented on aten/src/ATen/test/vec_test_all_types.h:
```diff
@@ -268,38 +269,42 @@ std::ostream& operator<<(std::ostream& stream, const CheckWithinDomains<T>& dmn)
 
 template<typename T>
-std::enable_if_t<!std::is_floating_point<T>::value, bool> check_both_big(T x, T y) {
+std::enable_if_t<!(std::is_floating_point<T>::value || std::is_reduced_floating_point_v<T>), bool> check_both_big(T x, T y) {
```
```suggestion
std::enable_if_t<!(std::is_floating_point_v<T>|| std::is_reduced_floating_point_v<T>), bool> check_both_big(T x, T y) {
```

malfet commented on aten/src/ATen/test/vec_test_all_types.h:
```diff
@@ -268,38 +269,42 @@ std::ostream& operator<<(std::ostream& stream, const CheckWithinDomains<T>& dmn)
 
 template <typename T>
-std::enable_if_t<!std::is_floating_point<T>::value, bool> check_both_inf(T x,
+std::enable_if_t<!(std::is_floating_point<T>::value || std::is_reduced_floating_point_v<T>), bool> check_both_inf(T x,
```
```suggestion
std::enable_if_t<!(std::is_floating_point_v<T>|| std::is_reduced_floating_point_v<T>), bool> check_both_inf(T x,
```

malfet commented on aten/src/ATen/test/vec_test_all_types.h:
```diff
@@ -268,38 +269,42 @@ std::ostream& operator<<(std::ostream& stream, const CheckWithinDomains<T>& dmn)
 
 template <typename T>
-std::enable_if_t<std::is_floating_point<T>::value, bool> check_both_inf(T x,
+std::enable_if_t<std::is_floating_point<T>::value || std::is_reduced_floating_point_v<T>, bool> check_both_inf(T x,
```
```suggestion
std::enable_if_t<std::is_floating_point_v<T>|| std::is_reduced_floating_point_v<T>, bool> check_both_inf(T x,
```

malfet commented on aten/src/ATen/test/vec_test_all_types.h:
```diff
@@ -268,38 +269,42 @@ std::ostream& operator<<(std::ostream& stream, const CheckWithinDomains<T>& dmn)
 
 template <typename T>
-std::enable_if_t<!std::is_floating_point<T>::value, bool> check_both_nan(T x,
+std::enable_if_t<!(std::is_floating_point<T>::value || std::is_reduced_floating_point_v<T>), bool> check_both_nan(T x,
```
```suggestion
std::enable_if_t<!(std::is_floating_point_v<T>|| std::is_reduced_floating_point_v<T>), bool> check_both_nan(T x,
```

malfet commented on aten/src/ATen/test/vec_test_all_types.h:
```diff
@@ -268,38 +269,42 @@ std::ostream& operator<<(std::ostream& stream, const CheckWithinDomains<T>& dmn)
 
 template <typename T>
-std::enable_if_t<std::is_floating_point<T>::value, bool> check_both_nan(T x,
+std::enable_if_t<std::is_floating_point<T>::value || std::is_reduced_floating_point_v<T>, bool> check_both_nan(T x,
```
```suggestion
std::enable_if_t<std::is_floating_point_v<T>|| std::is_reduced_floating_point_v<T>, bool> check_both_nan(T x,
```

malfet replied:
```suggestion
std::enable_if_t<std::is_floating_point_v<T>|| std::is_reduced_floating_point_v<T>, bool> check_both_nan(T x,
```

malfet commented on aten/src/ATen/test/vec_test_all_types.h:
```diff
@@ -15,6 +15,7 @@
 #include <math.h>
 #include <float.h>
 #include <algorithm>
+#include "c10/util/BFloat16-math.h"
```
```suggestion
```

metascroy committed:
add testing support for Vectorized<Half> (#122798)

Summary:

Implement testing support for Vectorized<Half>.

Test Plan:
Adjusted fbcode//caffe2:vec_test_all_types so that Half type is covered by tests.  On x86, this just tests the half type template in vec_base.h.  All tests pass.

```
buck2 test fbcode//caffe2:vec_test_all_types
Buck UI: https://www.internalfb.com/buck2/1062c33e-3e2b-4915-929d-5717f4d05ea8
Test UI: https://www.internalfb.com/intern/testinfra/testrun/15762598722963484
Network: Up: 7.6MiB  Down: 211MiB  (reSessionID-f19e6afc-3fa5-4455-87c5-df6f64d2ef70)
Jobs completed: 2872. Time elapsed: 8:17.6s.
Cache hits: 16%. Commands: 1571 (cached: 247, remote: 1321, local: 3)
Tests finished: Pass 336. Fail 0. Fatal 0. Skip 0. Build failure 0
```

Differential Revision: D55226003

malfet commented on aten/src/ATen/test/vec_test_all_types.cpp:
```diff
@@ -549,7 +557,8 @@ namespace {
         // double: 2e+305  float: 4e+36 (https://sleef.org/purec.xhtml#eg)
-        UVT maxCorrect = std::is_same<UVT, float>::value ? (UVT)4e+36 : (UVT)2e+305;
+        // Half not available in Sleef, but we set max at 4e+4.
+        UVT maxCorrect = std::is_same<UVT, c10::Half>::value ? (UVT)4e+4 : (std::is_same<UVT, float>::value ? (UVT)4e+36 : (UVT)2e+305);
```
```suggestion
        UVT maxCorrect = std::is_same_v<UVT, c10::Half> ? (UVT)4e+4 : (std::is_same_v<UVT, float> ? (UVT)4e+36 : (UVT)2e+305);
```

malfet committed:
Apply suggestions from code review

malfet committed:
Apply suggestions from code review

malfet committed:
Apply suggestions from code review

malfet committed:
Update Math.h
